File: app/src/main/java/kr/sofac/handsometalk/server/Connection.java
package kr.sofac.handsometalk.server;

import android.content.Context;
import android.net.Uri;

import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import com.google.gson.reflect.TypeToken;

import java.io.File;
import java.lang.reflect.Type;
import java.net.URISyntaxException;
import java.util.ArrayList;

import kr.sofac.handsometalk.dto.AuthorizationDTO;
import kr.sofac.handsometalk.dto.EstimateDTO;
import kr.sofac.handsometalk.dto.EventDTO;
import kr.sofac.handsometalk.dto.FeedbackDTO;
import kr.sofac.handsometalk.dto.GetEstimationsDTO;
import kr.sofac.handsometalk.dto.GetPushDTO;
import kr.sofac.handsometalk.dto.MessageDTO;
import kr.sofac.handsometalk.dto.NewEstimateRequestDTO;
import kr.sofac.handsometalk.dto.PushDTO;
import kr.sofac.handsometalk.dto.RegistrationDTO;
import kr.sofac.handsometalk.dto.UserDTO;
import kr.sofac.handsometalk.server.retrofit.ManagerRetrofit;
import kr.sofac.handsometalk.server.type.ServerResponse;
import kr.sofac.handsometalk.util.PathUtil;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import timber.log.Timber;


/**
 * Created by Maxim on 03.08.2017.
 */

public class Connection<T> {

    private AnswerServerResponse answerServerResponse;

    public interface AnswerServerResponse<T> {
        void processFinish(Boolean isSuccess, ServerResponse<T> answerServerResponse);
    }

    /**
     * Authorization DTO
     */
    public void authorizationUser(AuthorizationDTO authorizationDTO, AnswerServerResponse<T> async) { //Change name request / Change data in method parameters
        answerServerResponse = async;
        new ManagerRetrofit<AuthorizationDTO>().sendRequest(authorizationDTO, new Object() {// Change type Object sending / Change data sending
        }.getClass().getEnclosingMethod().getName(), (isSuccess, answerString) -> {
            if (isSuccess) {
                Type typeAnswer = new TypeToken<ServerResponse<UserDTO>>() { //Change type response
                }.getType();
                tryParsing(answerString, typeAnswer);
            } else {
                answerServerResponse.processFinish(false, null);
            }
        });
    }

    /**
     * Registration DTO
     */
    public void registrationUser(RegistrationDTO registrationDTO, AnswerServerResponse<T> async) { //Change name request / Change data in method parameters
        answerServerResponse = async;
        new ManagerRetrofit<RegistrationDTO>().sendRequest(registrationDTO, new Object() {// Change type Object sending / Change data sending
        }.getClass().getEnclosingMethod().getName(), (isSuccess, answerString) -> {
            if (isSuccess) {
                Type typeAnswer = new TypeToken<ServerResponse<UserDTO>>() { //Change type response
                }.getType();
                tryParsing(answerString, typeAnswer);
            } else {
                answerServerResponse.processFinish(false, null);
            }
        });
    }

    /**
     * Get All Events
     */
    public void allEvents(String str, AnswerServerResponse<T> async) { //Change name request / Change data in method parameters
        answerServerResponse = async;
        new ManagerRetrofit<String>().sendRequest(str, new Object() {// Change type Object sending / Change data sending
        }.getClass().getEnclosingMethod().getName(), (isSuccess, answerString) -> {
            if (isSuccess) {
                Type typeAnswer = new TypeToken<ServerResponse<ArrayList<EventDTO>>>() { //Change type response
                }.getType();
                tryParsing(answerString, typeAnswer);
            } else {
                answerServerResponse.processFinish(false, null);
            }
        });
    }

    /**
     * Get All Push notification
     */
    public void getListPush(GetPushDTO getPushDTO, AnswerServerResponse<T> async) { //Change name request / Change data in method parameters
        answerServerResponse = async;
        new ManagerRetrofit<GetPushDTO>().sendRequest(getPushDTO, new Object() {// Change type Object sending / Change data sending
        }.getClass().getEnclosingMethod().getName(), (isSuccess, answerString) -> {
            if (isSuccess) {
                Type typeAnswer = new TypeToken<ServerResponse<ArrayList<PushDTO>>>() { //Change type response
                }.getType();
                tryParsing(answerString, typeAnswer);
            } else {
                answerServerResponse.processFinish(false, null);
            }
        });
    }

    /**
     * Get All Estimations
     */
    public void getEstimations(GetEstimationsDTO getEstimationDTO, AnswerServerResponse<T> async) { //Change name request / Change data in method parameters
        answerServerResponse = async;
        new ManagerRetrofit<GetEstimationsDTO>().sendRequest(getEstimationDTO, new Object() {// Change type Object sending / Change data sending
        }.getClass().getEnclosingMethod().getName(), (isSuccess, answerString) -> {
            if (isSuccess) {
                Type typeAnswer = new TypeToken<ServerResponse<ArrayList<EstimateDTO>>>() { //Change type response
                }.getType();
                tryParsing(answerString, typeAnswer);
            } else {
                answerServerResponse.processFinish(false, null);
            }
        });
    }

    /**
     * Get Estimation and All Message inside
     */
    public void getEstimation(MessageDTO messageDTO, AnswerServerResponse<T> async) { //Change name request / Change data in method parameters
        answerServerResponse = async;
        new ManagerRetrofit<MessageDTO>().sendRequest(messageDTO, new Object() {// Change type Object sending / Change data sending
        }.getClass().getEnclosingMethod().getName(), (isSuccess, answerString) -> {
            if (isSuccess) {
                Type typeAnswer = new TypeToken<ServerResponse<EstimateDTO>>() { //Change type response
                }.getType();
                tryParsing(answerString, typeAnswer);
            } else {
                answerServerResponse.processFinish(false, null);
            }
        });
    }

    //{"dataTransferObject":{"estimate_id":"1"}, "requestType":"getEstimation"}

    public void newEstimation(Context context, NewEstimateRequestDTO newEstimateRequestDTO, ArrayList<Uri> listUri, AnswerServerResponse<T> async) {
        answerServerResponse = async;

        new ManagerRetrofit<NewEstimateRequestDTO>().sendMultiPartRequest(newEstimateRequestDTO, new Object() {// Change (type sending) / (data sending)
        }.getClass().getEnclosingMethod().getName(), generateMultiPartList(listUri, context), new ManagerRetrofit.AsyncAnswerString() {
            @Override
            public void processFinish(Boolean isSuccess, String answerString) {
                if (isSuccess) {
                    Type typeAnswer = new TypeToken<ServerResponse>() { //Change type response(тип ответа)
                    }.getType();
                    tryParsing(answerString, typeAnswer);
                } else {
                    answerServerResponse.processFinish(false, null);
                }
            }
        });

    }

    public void newMessage(Context context, MessageDTO messageDTO, ArrayList<Uri> listUri, AnswerServerResponse<T> async) {
        answerServerResponse = async;

        new ManagerRetrofit<MessageDTO>().sendMultiPartRequest(messageDTO, new Object() {// Change (type sending) / (data sending)
        }.getClass().getEnclosingMethod().getName(), generateMultiPartList(listUri, context), new ManagerRetrofit.AsyncAnswerString() {
            @Override
            public void processFinish(Boolean isSuccess, String answerString) {
                if (isSuccess) {
                    Type typeAnswer = new TypeToken<ServerResponse>() { //Change type response(тип ответа)
                    }.getType();
                    tryParsing(answerString, typeAnswer);
                } else {
                    answerServerResponse.processFinish(false, null);
                }
            }
        });

    }

    /**
     * Registration DTO
     */
    public void addFeedback(MessageDTO messageDTO, AnswerServerResponse<T> async) { //Change name request / Change data in method parameters
        answerServerResponse = async;
        new ManagerRetrofit<MessageDTO>().sendRequest(messageDTO, new Object() {// Change type Object sending / Change data sending
        }.getClass().getEnclosingMethod().getName(), (isSuccess, answerString) -> {
            if (isSuccess) {
                Type typeAnswer = new TypeToken<ServerResponse>() { //Change type response
                }.getType();
                tryParsing(answerString, typeAnswer);
            } else {
                answerServerResponse.processFinish(false, null);
            }
        });
    }

    /**
     * Registration DTO
     */
    public void feedbackList(GetEstimationsDTO getEstimationsDTO, AnswerServerResponse<T> async) { //Change name request / Change data in method parameters
        answerServerResponse = async;
        new ManagerRetrofit<GetEstimationsDTO>().sendRequest(getEstimationsDTO, new Object() {// Change type Object sending / Change data sending
        }.getClass().getEnclosingMethod().getName(), (isSuccess, answerString) -> {
            if (isSuccess) {
                Type typeAnswer = new TypeToken<ServerResponse<ArrayList<FeedbackDTO>>>() { //Change type response
                }.getType();
                tryParsing(answerString, typeAnswer);
            } else {
                answerServerResponse.processFinish(false, null);
            }
        });
    }

    /**
     * Registration DTO
     */
    public void updateUser(UserDTO userDTO, AnswerServerResponse<T> async) { //Change name request / Change data in method parameters
        answerServerResponse = async;
        new ManagerRetrofit<UserDTO>().sendRequest(userDTO, new Object() {// Change type Object sending / Change data sending
        }.getClass().getEnclosingMethod().getName(), (isSuccess, answerString) -> {
            if (isSuccess) {
                Type typeAnswer = new TypeToken<ServerResponse<ArrayList<FeedbackDTO>>>() { //Change type response
                }.getType();
                tryParsing(answerString, typeAnswer);
            } else {
                answerServerResponse.processFinish(false, null);
            }
        });
    }

    /**
     * //////////// Dop methods
     */

    public ArrayList<MultipartBody.Part> generateMultiPartList(ArrayList<Uri> listFileUri, Context context) {
        ArrayList<MultipartBody.Part> arrayListMulti = new ArrayList<>();
        for (int i = 0; i < listFileUri.size(); i++) {
            try {
                File file = new File(PathUtil.getPath(context, listFileUri.get(i)));
                arrayListMulti.add(MultipartBody.Part.createFormData("images[" + i + "]", file.getName(), RequestBody.create(MediaType.parse("multipart/form-data"), file)));
            } catch (URISyntaxException e) {
                e.printStackTrace();
            }

        }
        return arrayListMulti;
    }

    private void tryParsing(String answerString, Type typeAnswer) {
        try {
            answerServerResponse.processFinish(true, getObjectTransferFromJSON(answerString, typeAnswer));
        } catch (Exception e) {
            answerServerResponse.processFinish(false, null);
            e.printStackTrace();
        }
    }

    private ServerResponse<T> getObjectTransferFromJSON(String string, Type type) {
        try {
            return new Gson().fromJson(string, type);
        } catch (JsonSyntaxException e) {

            Timber.e("Не соответствующий тип данных для парсинга JSON");
            e.printStackTrace();
            return null;
        }
    }

}

File: app/src/main/java/kr/sofac/handsometalk/model/server/Connection.kt
package kr.sofac.handsometalk.model.server

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import kr.sofac.handsometalk.model.server.type.ServerResponse
import timber.log.Timber
import java.lang.reflect.Type


/**
 * Created by Maxim on 03.08.2017.
 */

class Connection<T> {

    private var answerServerResponse: AnswerServerResponse<*>? = null

    interface AnswerServerResponse<T> {
        fun processFinish(isSuccess: Boolean?, answerServerResponse: ServerResponse<T>?)
    }

    /**
     * Authorization DTO
     */
//    fun authorizationUser(authorizationDTO: AuthorizationDTO, async: AnswerServerResponse<T>) { //Change name request / Change data in method parameters
//        answerServerResponse = async
//        ManagerRetrofit<AuthorizationDTO>().sendRequest(authorizationDTO, object : Any() {// Change type Object sending / Change data sending
//
//        }.javaClass.enclosingMethod.name, { isSuccess, answerString ->
//            if (isSuccess!!) {
//                val typeAnswer = object : TypeToken<ServerResponse<UserDTO>>() { //Change type response
//
//                }.getType()
//                tryParsing(answerString, typeAnswer)
//            } else {
//                answerServerResponse!!.processFinish(false, null)
//            }
//        })
//    }
//
//    /**
//     * Get correct VERSION from Server
//     */
//    fun getCorrectVersion(async: AnswerServerResponse<T>) { //Change name request() / Change data in method parameters
//        answerServerResponse = async
//        ManagerRetrofit<String>().sendRequest("", object : Any() {// Change type Object sending / Change data sending
//
//        }.javaClass.enclosingMethod.name, { isSuccess, answerString ->
//            if (isSuccess!!) {
//                val typeAnswer = object : TypeToken<ServerResponse<AppVersionDTO>>() { //Change type response
//
//                }.getType()
//                tryParsing(answerString, typeAnswer)
//            } else {
//                answerServerResponse!!.processFinish(false, null)
//            }
//        })
//    }

    /**
     * Get List VERSIONS history from Server
     */
    fun getVersionsHistory(async: AnswerServerResponse<T>) {

    }

    /**
     * Get MANAGER info from server
     */
//    fun getManagerInfo(idManager: Long?, async: AnswerServerResponse<T>) {
//        answerServerResponse = async
//        ManagerRetrofit<Long>().sendRequest(idManager, object : Any() {// Change type Object sending / Change data sending
//
//        }.javaClass.enclosingMethod.name, { isSuccess, answerString ->
//            if (isSuccess!!) {
//                val typeAnswer = object : TypeToken<ServerResponse<ManagerDTO>>() { //Change type response
//
//                }.getType()
//                tryParsing(answerString, typeAnswer)
//            } else {
//                answerServerResponse!!.processFinish(false, null)
//            }
//        })
//    }

    /**
     * Get CUSTOMER info from server
     */
    fun getCustomerInfo(idCustomer: Long?, async: AnswerServerResponse<T>) {}

    /**
     * Get STAFF info from server
     */
    fun getStaffInfo(idStaff: Long?, async: AnswerServerResponse<T>) {}


    /**
     * Get SETTINGS from Server
     */
    fun getSettings() {

    }

//    fun getListComments(postId: Long?, async: AnswerServerResponse<T>) {
//        answerServerResponse = async
//        ManagerRetrofit<Long>().sendRequest(postId, object : Any() {// Change (type sending) / (data sending)
//
//        }.javaClass.enclosingMethod.name, { isSuccess, answerString ->
//            if (isSuccess!!) {
//                val typeAnswer = object : TypeToken<ServerResponse<ArrayList<CommentDTO>>>() { //Change type response(тип ответа)
//
//                }.getType()
//                tryParsing(answerString, typeAnswer)
//            } else {
//                answerServerResponse!!.processFinish(false, null)
//            }
//        })
//    }
//
//    /**    */
//    fun getListPosts(stringTypeGroup: String, async: AnswerServerResponse<T>) {
//        answerServerResponse = async
//        ManagerRetrofit<String>().sendRequest(stringTypeGroup, object : Any() {// Change (type sending) / (data sending)
//
//        }.javaClass.enclosingMethod.name, { isSuccess, answerString ->
//            if (isSuccess!!) {
//                val typeAnswer = object : TypeToken<ServerResponse<List<PostDTO>>>() { //Change type response(тип ответа)
//
//                }.getType()
//                tryParsing(answerString, typeAnswer)
//            } else {
//                answerServerResponse!!.processFinish(false, null)
//            }
//        })
//    }
//
//    /**    */
//    fun createPost(context: Context, postDTO: PostDTO, listUri: ArrayList<Uri>, async: AnswerServerResponse<T>) {
//        answerServerResponse = async
//
//        ManagerRetrofit<PostDTO>().sendMultiPartRequest(postDTO, object : Any() {// Change (type sending) / (data sending)
//
//        }.javaClass.enclosingMethod.name, generateMultiPartList(listUri, context), { isSuccess, answerString ->
//            if (isSuccess!!) {
//                val typeAnswer = object : TypeToken<ServerResponse<*>>() { //Change type response(тип ответа)
//
//                }.getType()
//                tryParsing(answerString, typeAnswer)
//            } else {
//                answerServerResponse!!.processFinish(false, null)
//            }
//        })
//
//    }
//
//
//    /**    */
//    fun updatePost(context: Context, postDTO: PostDTO, listUri: ArrayList<Uri>, listDeletingFiles: ArrayList<String>, async: AnswerServerResponse<T>) {
//        answerServerResponse = async
//
//        ManagerRetrofit<PostDTO>().sendMultiPartWithTwoObj(postDTO, object : Any() {// Change (type sending) / (data sending)
//
//        }.javaClass.enclosingMethod.name, generateMultiPartList(listUri, context), listDeletingFiles, { isSuccess, answerString ->
//            if (isSuccess!!) {
//                val typeAnswer = object : TypeToken<ServerResponse<*>>() { //Change type response(тип ответа)
//
//                }.getType()
//                tryParsing(answerString, typeAnswer)
//            } else {
//                answerServerResponse!!.processFinish(false, null)
//            }
//        })

        //        new ManagerRetrofit<PostDTO>().sendRequest(postDTO, new Object() {// Change (type sending) / (data sending)
        //        }.getClass().getEnclosingMethod().getName(), new ManagerRetrofit.AsyncAnswerString() {
        //            @Override
        //            public void processFinish(Boolean isSuccess, String answerString) {
        //                if (isSuccess) {
        //                    Type typeAnswer = new TypeToken<ServerResponse>() { //Change type response(тип ответа)
        //                    }.getType();
        //                    tryParsing(answerString, typeAnswer);
        //                } else {
        //                    answerServerResponse.processFinish(false, null);
        //                }
        //            }
        //        });
    //}

//    /**    */
//    fun deletePost(postDTO: PostDTO, async: AnswerServerResponse<T>) {
//        answerServerResponse = async
//        ManagerRetrofit<Long>().sendRequest(postDTO.getId(), object : Any() {// Change (type sending) / (data sending)
//
//        }.javaClass.enclosingMethod.name, { isSuccess, answerString ->
//            if (isSuccess!!) {
//                val typeAnswer = object : TypeToken<ServerResponse<*>>() { //Change type response(тип ответа)
//
//                }.getType()
//                tryParsing(answerString, typeAnswer)
//            } else {
//                answerServerResponse!!.processFinish(false, null)
//            }
//        })
//    }
//
//    /**
//     * COMMENT
//     */
//    fun createComment(commentDTO: CommentDTO, async: AnswerServerResponse<T>) {
//        answerServerResponse = async
//        ManagerRetrofit<CommentDTO>().sendRequest(commentDTO, object : Any() {// Change (type sending) / (data sending)
//
//        }.javaClass.enclosingMethod.name, { isSuccess, answerString ->
//            if (isSuccess!!) {
//                val typeAnswer = object : TypeToken<ServerResponse<*>>() { //Change type response(тип ответа)
//
//                }.getType()
//                tryParsing(answerString, typeAnswer)
//            } else {
//                answerServerResponse!!.processFinish(false, null)
//            }
//        })
//    }
//
//    /**  */
//    fun updateComment(commentDTO: CommentDTO, async: AnswerServerResponse<T>) {
//        answerServerResponse = async
//        ManagerRetrofit<CommentDTO>().sendRequest(commentDTO, object : Any() {// Change (type sending) / (data sending)
//
//        }.javaClass.enclosingMethod.name, { isSuccess, answerString ->
//            if (isSuccess!!) {
//                val typeAnswer = object : TypeToken<ServerResponse<*>>() { //Change type response(тип ответа)
//
//                }.getType()
//                tryParsing(answerString, typeAnswer)
//            } else {
//                answerServerResponse!!.processFinish(false, null)
//            }
//        })
//    }
//
//    /**    */
//    fun deleteComment(commentDTO: CommentDTO, async: AnswerServerResponse<T>) {
//        answerServerResponse = async
//        ManagerRetrofit<Long>().sendRequest(commentDTO.getId(), object : Any() {// Change (type sending) / (data sending)
//
//        }.javaClass.enclosingMethod.name, { isSuccess, answerString ->
//            if (isSuccess!!) {
//                val typeAnswer = object : TypeToken<ServerResponse<*>>() { //Change type response(тип ответа)
//
//                }.getType()
//                tryParsing(answerString, typeAnswer)
//            } else {
//                answerServerResponse!!.processFinish(false, null)
//            }
//        })
//    }


    /**
     * Dop methods
     */

//    fun generateMultiPartList(listFileUri: ArrayList<Uri>, context: Context): ArrayList<MultipartBody.Part> {
//        val arrayListMulti = ArrayList<MultipartBody.Part>()
//        for (i in listFileUri.indices) {
//            Timber.e("listFileUri.get(i).toString()  " + listFileUri[i].toString())
//            try {
//                val file = File(PathUtil.getPath(context, listFileUri[i]))
//                arrayListMulti.add(MultipartBody.Part.createFormData("files[$i]", file.getName(), RequestBody.create(MediaType.parse("multipart/form-data"), file)))
//            } catch (e: URISyntaxException) {
//                e.printStackTrace()
//            }
//
//        }
//        return arrayListMulti
//    }

    private fun tryParsing(answerString: String, typeAnswer: Type) {
        try {
            answerServerResponse!!.processFinish(true, getObjectTransferFromJSON(answerString, typeAnswer))
        } catch (e: Exception) {
            answerServerResponse!!.processFinish(false, null)
            e.printStackTrace()
        }

    }

    private fun getObjectTransferFromJSON(string: String, type: Type): ServerResponse<T>? {
        try {
            return Gson().fromJson(string, type)
        } catch (e: JsonSyntaxException) {

            Timber.e("Не соответствующий тип данных для парсинга JSON")
            e.printStackTrace()
            return null
        }

    }

}
